### ISSUE TITLE
About the image_processor

### ISSUE THREAD

**Issue by dszpr** [OpenGVLab/InternVL#357]
Hi, really appreciate for your amazing work!
In LLaVA, the image_processor is CLIPImageProcessor. However, I can't locate the CLASS or FUNCTION of the image_processor when using Intern-ViT. Specifically, the image_processor is used in the train.py. I wonder whether the the image_processor is same in LLaVA and InternVL?
Looking forward to your reply!
`
    def __getitem__(self, i) -> Dict[str, torch.Tensor]:
        flag = False
        while not flag:
            try:
                sources = self.list_data_dict[i]
                if isinstance(i, int):
                    sources = [sources]
                assert len(sources) == 1, "Don't know why it is wrapped to a list"  # FIXME
                if 'image' in sources[0]:
                    image_file = self.list_data_dict[i]['image']
                    image_folder = self.data_args.image_folder
                    processor = self.data_args.image_processor
                    image = Image.open(os.path.join(image_folder, image_file)).convert('RGB')
                    if self.data_args.image_aspect_ratio == 'pad':
                        def expand2square(pil_img, background_color):
                            width, height = pil_img.size
                            if width == height:
                                return pil_img
                            elif width > height:
                                result = Image.new(pil_img.mode, (width, width), background_color)
                                result.paste(pil_img, (0, (width - height) // 2))
                                return result
                            else:
                                result = Image.new(pil_img.mode, (height, height), background_color)
                                result.paste(pil_img, ((height - width) // 2, 0))
                                return result
                        image = expand2square(image, tuple(int(x*255) for x in processor.image_mean))
                        image = processor.preprocess(image, return_tensors='pt')['pixel_values'][0]
                    else:
                        image = processor.preprocess(image, return_tensors='pt')['pixel_values'][0]
                    sources = preprocess_multimodal(
                        copy.deepcopy([e["conversations"] for e in sources]),
                        self.data_args)

`

**Comment by G-z-w** [comment]
> Hi, really appreciate for your amazing work! In LLaVA, the image_processor is CLIPImageProcessor. However, I can't locate the CLASS or FUNCTION of the image_processor when using Intern-ViT. Specifically, the image_processor is used in the train.py. I wonder whether the the image_processor is same in LLaVA and InternVL? Looking forward to your reply! ` def **getitem**(self, i) -> Dict[str, torch.Tensor]: flag = False while not flag: try: sources = self.list_data_dict[i] if isinstance(i, int): sources = [sources] assert len(sources) == 1, "Don't know why it is wrapped to a list" # FIXME if 'image' in sources[0]: image_file = self.list_data_dict[i]['image'] image_folder = self.data_args.image_folder processor = self.data_args.image_processor image = Image.open(os.path.join(image_folder, image_file)).convert('RGB') if self.data_args.image_aspect_ratio == 'pad': def expand2square(pil_img, background_color): width, height = pil_img.size if width == height: return pil_img elif width > height: result = Image.new(pil_img.mode, (width, width), background_color) result.paste(pil_img, (0, (width - height) // 2)) return result else: result = Image.new(pil_img.mode, (height, height), background_color) result.paste(pil_img, ((height - width) // 2, 0)) return result image = expand2square(image, tuple(int(x*255) for x in processor.image_mean)) image = processor.preprocess(image, return_tensors='pt')['pixel_values'][0] else: image = processor.preprocess(image, return_tensors='pt')['pixel_values'][0] sources = preprocess_multimodal( copy.deepcopy([e["conversations"] for e in sources]), self.data_args)
> 
> `

Hi, the image transform we used is In [code](https://github.com/OpenGVLab/InternVL/blob/53fdd1fdc64fce5fa20db2f7779479500fcb961c/internvl_chat/internvl/train/dataset.py#L92), It does the same thing as CLIPImageProcessor, and you can compare the parameters of the two in detail.

**Comment by dszpr** [comment]
CLIPImageProcessor returns pixel value. I notice that in the example inference code of InterVL2, the image process code of dynamic resolution is provided. I wonder whether it work if I replace CLIPImageProcessor with your image process code of dynamic resolution? 
`
import math
import numpy as np
import torch
import torchvision.transforms as T
from decord import VideoReader, cpu
from PIL import Image
from torchvision.transforms.functional import InterpolationMode
from transformers import AutoModel, AutoTokenizer

IMAGENET_MEAN = (0.485, 0.456, 0.406)
IMAGENET_STD = (0.229, 0.224, 0.225)


def build_transform(input_size):
    MEAN, STD = IMAGENET_MEAN, IMAGENET_STD
    transform = T.Compose([
        T.Lambda(lambda img: img.convert('RGB') if img.mode != 'RGB' else img),
        T.Resize((input_size, input_size), interpolation=InterpolationMode.BICUBIC),
        T.ToTensor(),
        T.Normalize(mean=MEAN, std=STD)
    ])
    return transform


def find_closest_aspect_ratio(aspect_ratio, target_ratios, width, height, image_size):
    best_ratio_diff = float('inf')
    best_ratio = (1, 1)
    area = width * height
    for ratio in target_ratios:
        target_aspect_ratio = ratio[0] / ratio[1]
        ratio_diff = abs(aspect_ratio - target_aspect_ratio)
        if ratio_diff < best_ratio_diff:
            best_ratio_diff = ratio_diff
            best_ratio = ratio
        elif ratio_diff == best_ratio_diff:
            if area > 0.5 * image_size * image_size * ratio[0] * ratio[1]:
                best_ratio = ratio
    return best_ratio


def dynamic_preprocess(image, min_num=1, max_num=6, image_size=448, use_thumbnail=False):
    orig_width, orig_height = image.size
    aspect_ratio = orig_width / orig_height

    # calculate the existing image aspect ratio
    target_ratios = set(
        (i, j) for n in range(min_num, max_num + 1) for i in range(1, n + 1) for j in range(1, n + 1) if
        i * j <= max_num and i * j >= min_num)
    target_ratios = sorted(target_ratios, key=lambda x: x[0] * x[1])

    # find the closest aspect ratio to the target
    target_aspect_ratio = find_closest_aspect_ratio(
        aspect_ratio, target_ratios, orig_width, orig_height, image_size)

    # calculate the target width and height
    target_width = image_size * target_aspect_ratio[0]
    target_height = image_size * target_aspect_ratio[1]
    blocks = target_aspect_ratio[0] * target_aspect_ratio[1]

    # resize the image
    resized_img = image.resize((target_width, target_height))
    processed_images = []
    for i in range(blocks):
        box = (
            (i % (target_width // image_size)) * image_size,
            (i // (target_width // image_size)) * image_size,
            ((i % (target_width // image_size)) + 1) * image_size,
            ((i // (target_width // image_size)) + 1) * image_size
        )
        # split the image
        split_img = resized_img.crop(box)
        processed_images.append(split_img)
    assert len(processed_images) == blocks
    if use_thumbnail and len(processed_images) != 1:
        thumbnail_img = image.resize((image_size, image_size))
        processed_images.append(thumbnail_img)
    return processed_images


def load_image(image_file, input_size=448, max_num=6):
    image = Image.open(image_file).convert('RGB')
    transform = build_transform(input_size=input_size)
    images = dynamic_preprocess(image, image_size=input_size, use_thumbnail=True, max_num=max_num)
    pixel_values = [transform(image) for image in images]
    pixel_values = torch.stack(pixel_values)
    return pixel_values


`

**Comment by G-z-w** [comment]
If you are only using internvit, you do not need to use dynamic resolution. Dynamic resolution is used to enable internvl to support larger resolutions by dividing the image into several tiles. Internvit will process each tile as an independent image, and then the embeddings of each tile will be aggregated and input into the LLM.

**Comment by dszpr** [comment]
Indeed, I am finetuning InternVL2-26B on my own task. However, I don't find the training code of InternVL2-26B and I need to modify llava training code to run finetuning. That is why I want to figure out the image_processor. 
Not sure if I understand this correctly：if I want to finetune InternVL2, dynamic resolution is necessary in the training code, or if I only use internvit as VE with other LLM, dynamic resolution is unnecessary.

**Comment by G-z-w** [comment]
The dynamic resolution is necessary only in our training code. You can use dynamic resolution and our training code to make other LLM support higher resolutions, and we will release our InternVL2 finetune code soon.

**Comment by dszpr** [comment]
Got it. Thanks again and looking forward to your finetune code!
